text/0000-fabric-x-next-generation.md

```diff
@@ -0,0 +1,365 @@
+* **End-to-End (E2E) Scenario Testing:** Simulating high-volume use cases, concurrent submissions from multiple clients, data consistency checks across replicas of the distributed database, and failure recovery of various microservices.
+* **Security Testing:** Penetration testing of individual microservices and their APIs, ensuring secure inter-service communication (e.g., via mTLS), validation of threshold signature implementation against known attacks, and ensuring the pre-stored identity mechanism is secure.
+* **Long-Running Stability Tests:** Operating the system under sustained load for extended periods to identify memory leaks, resource exhaustion, performance degradation over time, and ensure overall system stability.
+
+**Integration Test Scenarios (Examples):**
+1.  Successful Transaction Flow: Client -> Endorser (via SDK/Flow) -> Arma Orderer (Router, Batcher, Consenter, Assembler) -> Sidecar -> Coordinator -> Signature Verifier -> VC -> Ledger. Verify state update in the distributed database and Sidecar event notification to the client.
+2.  Transaction with Invalid Endorsement: Ensure rejection at the Signature Verifier or VC Service, and that the transaction is marked invalid but still included in the block (as per Fabric's model).
+3.  Conflicting Transactions in Parallel Validation: Submit a mix of dependent and independent transactions. Verify that the dependency graph correctly identifies conflicts, independent transactions are processed in parallel, and dependent transactions are processed in the correct order or invalidated as appropriate.
+4.  Arma Primary Batcher Failure: Simulate a primary batcher crash or Byzantine behavior in a shard. Verify that secondary batchers detect this, complaints are processed by consenters, a new primary is elected, and the shard continues to process transactions.
+5.  Sidecar Fails to Pull Blocks: Simulate Sidecar failure. Upon restart, verify it can resume pulling blocks from Arma from the correct point and correctly forward them to the Coordinator.
+6.  Distributed Database Node Failure: Simulate failure of one or more nodes in the organization's distributed database cluster. Verify that the database maintains consistency and availability (as per its design, e.g., if Raft-based, it should tolerate F failures in a 2F+1 cluster) and that VC services can continue to operate.
+7.  Horizontal Scaling Test: Incrementally add instances of Endorsers, Signature Verifiers, and VC Services. Measure the impact on throughput and latency to confirm effective scaling.
+
+# Dependencies
+[dependencies]: #dependencies
+
+* **External Libraries/Frameworks:**
+    * **SmartBFT Library (or equivalent BFT library):** For Arma's consensus mechanism.
+    * **Distributed Database System:** A specific choice will need to be integrated (e.g., YugabyteDB, CockroachDB, TiDB, or potentially a pluggable interface).
+    * **gRPC/Messaging System:** For efficient and reliable inter-microservice communication.
```

Comment:
>**gRPC/Messaging System:** For efficient and reliable inter-microservice communication.

It would be helpful to properly manage this in a separate repository, away from other code/common utilities. I think Fabric extracted a `fabric-protos` repo quite late on and I think there are probably questionable things in there that only happened because the API had previously been treated no differently to other code changes.

Reply:
Yes, We will do that.